Zend/zend_partial.c

```diff
@@ -0,0 +1,589 @@
+
+	ZEND_PARSE_PARAMETERS_START(0, -1)
+		Z_PARAM_VARIADIC_WITH_NAMED(params, num_params, named_args)
+	ZEND_PARSE_PARAMETERS_END();
+
+	fci.size = sizeof(zend_fcall_info);
+	fci.retval = return_value;
+	fci.param_count = MAX(partial->func.common.num_args, partial->argc + num_params);
+	fci.params = ecalloc(sizeof(zval), fci.param_count);
+	fcc.function_handler = &partial->func;
+
+	if (Z_TYPE(partial->This) == IS_OBJECT) {
+	    object = zend_partial_this(Z_OBJ(partial->This));
+	} else if (Z_TYPE(partial->This) == IS_PTR) {
+	    zend_class_entry *type = Z_PTR(partial->This);
+	    zval instance;
+
+        if (type->create_object) {
+            ZVAL_OBJ(&instance, type->create_object(type));
+        } else {
+            ZVAL_OBJ(&instance, zend_objects_new(type));
+        }
+        
+        object = Z_OBJ(instance);
+        
+        GC_ADD_FLAGS(object, IS_OBJ_DESTRUCTOR_CALLED);
+	}
+	
+	if (object) {
+	    fci.object = object;
+	    fcc.object = object;
+	}
+
+	memcpy(fci.params, partial->argv, partial->argc * sizeof(zval));
+
+    zval *arg = fci.params,
+         *aend = arg + fci.param_count,
+         *pend = params + num_params;
+
+    while (arg < aend) {
+        if (Z_TYPE_P(arg) == IS_UNDEF) {
+            if (params < pend) {
+                ZVAL_COPY_VALUE(arg, params);
+                params++;
+            } else {
+                fci.param_count--;
+            }
+        }
+
+        arg++;
+    }
+
+    if (partial->named) {
+        if (!named_args) {
+            named_args = partial->named;
+            GC_ADDREF(named_args);
+        } else {
+            HashTable *nested = zend_array_dup(partial->named);
+            
+            zend_hash_merge(nested, named_args, zval_copy_ctor, true);
+            
+            named_args = nested;
+        }
+        fci.named_params = named_args;
+    } else {
+        fci.named_params = named_args;
+    }
+
+    zend_call_function(&fci, &fcc);
+
+    if (!EG(exception)) {
+        if (Z_TYPE(partial->This) == IS_PTR) {
+            GC_DEL_FLAGS(object, IS_OBJ_DESTRUCTOR_CALLED);
+        }
+    }
+
+    if (partial->named) {
+        zend_array_release(named_args);
+    }
+
+    if (partial->func.common.fn_flags & ZEND_ACC_CTOR) {
+        zval_ptr_dtor(return_value);
+
+        RETVAL_OBJ(object);
+    }
+    
+    if (!fcc.function_handler) {
+        EG(trampoline).common.function_name = NULL;
+    }
+    
+    efree(fci.params);
+}
+
+void zend_partial_apply(zval *result, uint32_t info, zend_execute_data *call) {
+    zend_function *function, *prototype = NULL;
+
+    ZVAL_OBJ(result, zend_partial_new(zend_ce_closure));
+
+    zend_partial *parent, *partial = (zend_partial*) Z_OBJ_P(result);
+
+    if ((parent = zend_partial_parent(&call->This))) {
+        function  = &parent->func;
+        prototype = &parent->prototype;
+
+        partial->argc = parent->argc + ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
```

Comment:
Can emalloc, no point in zeroing if you're overwriting it in the next line.

Reply:
This is setting undef unused slots, it's better to use calloc.

Reply:
You're right, I didn't notice that one is partial->argc while the other is parent->argc.